adam_core/orbits/variants.py

```diff
@@ -0,0 +1,177 @@
+        # once the changes have been merged
+        return qv.MultiKeyLinkage(
+            orbits,
+            self,
+            left_keys={
+                "orbit_id": orbits.orbit_id,
+                "jd1": orbits.coordinates.time.jd1,
+                "jd2": orbits.coordinates.time.jd2,
+            },
+            right_keys={
+                "orbit_id": self.orbit_id,
+                "jd1": self.coordinates.time.jd1,
+                "jd2": self.coordinates.time.jd2,
+            },
+        )
+
+    def collapse(self, orbits: Orbits) -> Orbits:
+        """
+        Collapse the variants and recalculate the covariance matrix for each
+        each orbit at each epoch. The mean state is taken from the orbits class and
+        is not calculate from the variants.
+
+        Parameters
+        ----------
+        orbits : `~adam_core.orbits.orbits.Orbits`
+            Orbits from which the variants were generated.
+
+        Returns
+        -------
+        collapsed_orbits : `~adam_core.orbits.orbits.Orbits`
+            The collapsed orbits.
+        """
+        link = self.link_to_orbits(orbits)
+
+        # Iterate over the variants and calculate the mean state and covariance matrix
+        # for each orbit at each epoch then create a new orbit with the calculated covariance matrix
+        orbits_list = []
+        for key, orbit, variants in link.iterate():
+            key = key.as_py()
+
+            assert len(orbit) == 1
+
+            samples = variants.coordinates.values
+            mean_state, covariance = mean_and_covariance_from_weighted_samples(
+                samples, variants.weights.to_numpy(), variants.weights_cov.to_numpy()
+            )
+
+            orbit_collapsed = Orbits.from_kwargs(
+                orbit_id=orbit.orbit_id,
+                object_id=orbit.object_id,
+                coordinates=CartesianCoordinates.from_kwargs(
+                    time=orbit.coordinates.time,
+                    x=orbit.coordinates.x,
+                    y=orbit.coordinates.y,
+                    z=orbit.coordinates.z,
+                    vx=orbit.coordinates.vx,
+                    vy=orbit.coordinates.vy,
+                    vz=orbit.coordinates.vz,
+                    covariance=CoordinateCovariances.from_matrix(covariance),
+                    frame=orbit.coordinates.frame,
+                    origin=orbit.coordinates.origin,
+                ),
+            )
+
+            orbits_list.append(orbit_collapsed)
+
+        orbits_collapsed = qv.concatenate(orbits_list)
+
+        # Array of indices into the collapsed orbits
+        orbits_idx = pa.array(np.arange(0, len(orbits_collapsed)))
+
+        # Make a list of arrays that will be used to sort the orbits
+        orbits_idx_sorted_list = []
+
+        # Loop over input orbits and figure out where in the collapsed orbits they occur
+        # There has to be an easier or better way to do this?
```

Comment:
> There has to be an easier or better way to do this?

Yes, there is! Instead of using `link.iterate()`, construct your own keys and query in the order you choose. Something like this:

```py
for orbit in orbits:
    key = link.key(
        orbit_id=orbits.orbit_id[0],
        jd1=orbits.coordinates.time.jd1[0],
        jd2=orbits.coordinates.time.jd2[0],
    )
    variants = link.select_right(key)
    # As above, now:
    coords = orbit.coordinates
    coords.covariance = CoordinateCovariances.from_matrix(variants.weighted_covariance())
    orbit.coordinates = coordinates
    orbits_list.append(orbit)
return qv.concatenate(orbits_list)    
```

Reply:
Sorted iteration like this is super nice! Implemented.